### PR TITLE
Cmake fixes; fix so library compilation error

### PR DESCRIPTION
**Pull request by Warchant**
- use STATIC target for tests, use sources for libed25519.so
- use exportmap to filter global methods in SO, this fixes compilation
  error of ed25519  https://github.com/hyperledger/iroha-ed25519/issues/6
- move test/ed25519/tosha3 to util/tosha3
- to simplify CI pipeline, approach "build what you use" was cancelled.
  Currently all targets are available by default (so you can write make <target>) but excluded from 'all' (i.e. make all will not build certain targets)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,20 +1,31 @@
-cmake_minimum_required(VERSION 3.0)
+cmake_minimum_required(VERSION 3.5)
```
Comment (Warchant) [inline CMakeLists.txt]: I was forced to update minimal version because of  `cmake_parse_arguments`, which was introduced in 3.5.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,20 +1,31 @@
...
-SET(CMAKE_EXPORT_COMPILE_COMMANDS "ON")
+
+set(CMAKE_POSITION_INDEPENDENT_CODE    TRUE)
+set(CMAKE_EXPORT_COMPILE_COMMANDS      ON)
+set(CMAKE_MODULE_PATH                  "${CMAKE_CURRENT_SOURCE_DIR}/cmake/Modules")
+set(EXECUTABLE_OUTPUT_PATH             "${PROJECT_BINARY_DIR}/bin")
+
+set(CMAKE_C_STANDARD                   11)    # force std=gnu11
+set(CMAKE_C_STANDARD_REQUIRED          ON)
+set(CMAKE_C_EXTENSIONS                 OFF)
```
Comment (l4l) [inline CMakeLists.txt]: why is that here

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,20 +1,31 @@
...
+
+set(CMAKE_CXX_STANDARD                 11)    # force std=c++11
+set(CMAKE_CXX_STANDARD_REQUIRED        ON)
+set(CMAKE_CXX_EXTENSIONS               OFF)
+set(CMAKE_CXX_FLAGS_RELEASE            "${CMAKE_CXX_FLAGS_RELEASE} -O3 -NDEBUG")
+set(CMAKE_CXX_FLAGS_DEBUG              "${CMAKE_CXX_FLAGS_DEBUG} -g -O0 -fdiagnostics-color")
+
+add_compile_options(-Wall)
+add_compile_options(-Wextra)
+add_compile_options(-Wpedantic)
```
Comment (l4l) [inline CMakeLists.txt]: `-Werror` maybe as well?

File: README.md [hunk 4]
```diff
@@ -20,7 +20,7 @@ During CMake time, users are able to choose any of these implementations using c
 
 - `EDIMPL`
     - `ref10` - portable C implementation. 
-    - `amd64-64-24k` - optimized C and ASM implementation, works only on Linux amd64. 
+    - `amd64-64-24k` - optimized C and ASM implementation, works only on Linux amd64. **This option is likely to fail during compilation of `ed25519` library, use `amd64-64-24k-pic`.
```
Comment (l4l) [inline README.md]: maybe just remove that target then?

File: test/hash/sha_test.cpp [hunk 5]
```diff
@@ -1,7 +1,7 @@
 #include <ed25519/ed25519/sha256.h>
 #include <ed25519/ed25519/sha512.h>
 #include <gtest/gtest.h>
-#include "hexutil.hpp"
+#include "../../util/hexutil.hpp"
```
Comment (l4l) [inline test/hash/sha_test.cpp]: :( add include dir

File: util/helpers.hpp [hunk 6]
```diff
@@ -1,11 +1,11 @@
 #ifndef ED25519_HELPERS_HPP_
 #define ED25519_HELPERS_HPP_
 
-#include <fstream>
-#include <sstream>
-#include <string>
-#include <vector>
-#include <ed25519/ed25519.h>
+#include "../../../../../usr/include/c++/5/fstream"
```
Comment (l4l) [inline util/helpers.hpp]: D: orly?
I have no such file

File: README.md (hunk 4 above)
Comment (Warchant) [reply README.md]: This target can be used in STATIC build (theoretically). To test this I need a ci.

File: CMakeLists.txt (hunk 3 above)
Comment (Warchant) [reply CMakeLists.txt]: Build will fail for most of targets and tests. I will fix all warnings as soon as CI will  be completed.

File: util/helpers.hpp (hunk 6 above)
Comment (Warchant) [reply util/helpers.hpp]: this is clion... :(

File: CMakeLists.txt (hunk 2 above)
Comment (Warchant) [reply CMakeLists.txt]: https://cmake.org/cmake/help/v3.3/prop_tgt/C_EXTENSIONS.html#prop_tgt:C_EXTENSIONS
